### PR TITLE
General refactoring of the OculusModule

### LINKED ISSUE #3: Take into account joint limits in fingers retargeting
The joint limits should be take into account for the finger retargeting. Indeed the desired position of the fingers are evaluated using an integrator 
https://github.com/robotology/walking-teleoperation/blob/515abfe571432b88610cd344c7c41082184294fe/modules/Oculus_module/src/FingersRetargeting.cpp#L47
Since the integrator is not saturated the desired joint position can exceed the limits.

I think that this can be easily fixed by setting the limits to the integrator object. Please check [here](http://www.icub.org/software_documentation/classiCub_1_1ctrl_1_1Integrator.html#a416b570ee6038b25f7706bd781d69b20) for further details 
https://github.com/robotology/walking-teleoperation/blob/515abfe571432b88610cd344c7c41082184294fe/modules/Oculus_module/src/FingersRetargeting.cpp#L34

### LINKED ISSUE #4: Use device interfaces instead of ports for getting input from oculus
It would be nice to use the oculus YARP [device](http://www.yarp.it/group__dev__iface.html) for getting the inputs instead of using YARP ports. 
https://github.com/robotology/walking-teleoperation/blob/515abfe571432b88610cd344c7c41082184294fe/modules/Oculus_module/src/OculusModule.cpp#L557-L564

### LINKED ISSUE #5: Map Oculus buttons into defined YARP commands
It would be nice to map the joypad oculus buttons to defined YARP commands. E.g.
* A + B `calib all 300`
* X + Y `startWalking`
* A + X reset the inertial frame of the robot. 
* ...

In my opinion, this may simplify the setup of the demo.

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
With this PR:
1. the code related to the Oculus is completely rewritten. The idea is to increase the modularity and prevent the stopping of the demo if a robot hardware issue occurs (i.e., one of the fingers goes in hardware fault) 
2.  now it is possible to run the entire walking demo directly using the joypad buttons. In details, the `A` button is used to prepare the robot and `X` to start walking (Close #5 )
3. The grasping feature is now implemented using the analog button of the oculus joypad. As a consequence, you can control the velocity of the fingers. 
4. fingers limits are taken into account in retargeting (Close #3 )
5. [`JoypadControlClient`](http://www.yarp.it/classyarp_1_1dev_1_1JoypadControlClient.html), instead of YARP ports,  is used to get the buttons and axis signal from the oculus  joypad (Close #4);
6. The headset orientation can be retrieved using either `TransformClient` (a feature not implemented in `ovrheadset` device https://github.com/robotology/yarp/issues/1937) or a YARP port.

TODO:
- [x] Improve the code documentation
- [x] test on iCub

cc @kouroshD @s-dafarra @DanielePucci @traversaro

### CODE REVIEW

File: app/robots/iCubGenova04/.#leftFingersRetargetingParams.ini [hunk 1]
```diff
@@ -0,0 +1 @@
+icub@icub-head.2463:1545383651
```
Comment (traversaro) [inline app/robots/iCubGenova04/.#leftFingersRetargetingParams.ini]: emacs leftover?

File: app/robots/iCubGenova04/headRetargetingParams.ini [hunk 2]
```diff
@@ -1 +1,6 @@
-smoothingTime   0.5
+remote_control_boards   ("head")
+# Notice the order of the joint list is not wrong.
```
Comment (traversaro) [inline app/robots/iCubGenova04/headRetargetingParams.ini]: This is not superclear to  be honest

File: app/robots/iCubGenova04/.#leftFingersRetargetingParams.ini (hunk 1 above)
Comment (GiulioRomualdi) [reply app/robots/iCubGenova04/.#leftFingersRetargetingParams.ini]: Ops. I have to update `.gitignore`!

File: modules/Oculus_module/CMakeLists.txt [hunk 3]
```diff
@@ -18,6 +18,7 @@ add_install_rpath_support(BIN_DIRS "${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_BIND
 find_package(ICUB REQUIRED)
 find_package(Eigen3 REQUIRED)
 find_package(iDynTree REQUIRED)
+find_package(OpenCV REQUIRED)
```
Comment (traversaro) [inline modules/Oculus_module/CMakeLists.txt]: Why is this necessary? I see no new code that uses OpenCV.

File: modules/Oculus_module/src/OculusModule.cpp [hunk 4]
```diff
@@ -229,7 +183,7 @@ bool OculusModule::configure(yarp::os::ResourceFinder& rf)
     // check if the configuration file is empty
     if (rf.isNull())
     {
-        yError() << "[configure] Empty configuration for the force torque sensors.";
+        yError() << "[OculusModule::configure] Empty configuration for the force torque sensors.";
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/OculusModule.cpp]: FT sensors?

File: modules/Oculus_module/src/FingersRetargeting.cpp [hunk 5]
```diff
@@ -7,63 +7,58 @@
...
+        samplingTime, buff, m_controlHelper->getLimits());
 
-bool FingersRetargeting::closeHand()
-{
-    if (m_fingerIntegrator == nullptr)
+    if (!m_controlHelper->switchToControlMode(VOCAB_CM_POSITION_DIRECT))
     {
-        yError() << "[closeHand] the integrator is not initialized please call "
-                    "configure() method";
+        yError() << "unable to switch the control mode";
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/FingersRetargeting.cpp]: Maybe we can avoid to ``return false`` here since some joints may be in HF. We may not care too much if some joints won't move in this case.

File: modules/Oculus_module/src/RobotControlHelper.cpp [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+    for (int i = 0; i < m_actuatedDOFs; i++)
+    {
+        m_desiredPositionInDegrees(i) = iDynTree::rad2deg(desiredPosition(i));
+
+        // check if the position is outside the limit
+        if (m_desiredPositionInDegrees(i) <= m_minJointsPosition(i))
+            m_desiredPositionInDegrees(i) = m_minJointsPosition(i) + threshold;
+        else if (m_desiredPositionInDegrees(i) >= m_maxJointsPosition(i))
+            m_desiredPositionInDegrees(i) = m_maxJointsPosition(i) - threshold;
+    }
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/RobotControlHelper.cpp]: The references should be already clamped at firmware level. In this way you don't need to resort to the limits in this method (i.e. it is not a problem if some joints are in the "not configured" status).

File: modules/Oculus_module/src/RobotControlHelper.cpp [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+    if (!okPosition)
+    {
+        yError() << "[configure] Unable to read encoders (position).";
+        m_isActive = false;
+        return false;
+    }
+
+    for (int i = 0; i < m_actuatedDOFs; i++)
+        // get position limits
+        if (!m_limitsInterface->getLimits(i, &m_minJointsPosition(i), &m_maxJointsPosition(i)))
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/RobotControlHelper.cpp]: I would get the limits in the ``getLimits`` method. In this way the ``configure`` phase does not fail if some joints are in the "not configured" status. Also, we can save a vector of bools to understand which joint is currently not working.

File: modules/Oculus_module/src/RobotControlHelper.cpp [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+}
+
+int RobotControlHelper::getDoFs()
+{
+    return m_actuatedDOFs;
+}
+
+yarp::sig::Matrix RobotControlHelper::getLimits()
+{
+    yarp::sig::Matrix limits(m_minJointsPosition.size(), 2);
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/RobotControlHelper.cpp]: I would read the joint limits here. In case it is not possible to retrieve a limit, simply set it equal to zero on both sides. Eventually you can also output a vector of bools specifying which joint is in a "not configured" status.

File: modules/Oculus_module/src/OculusModule.cpp [hunk 9]
```diff
@@ -496,230 +326,175 @@ double OculusModule::getPeriod()
...
+        yError() << "[OculusModule::updateModule] Unable to set the right finger velocity.";
+        return false;
     }
-
-    // move the robot
-    if (!setDirectPositionReferences(m_qDesired))
+    if (!m_rightHandFingers->move())
     {
-        yError() << "[updateModule] Unable to move the robot";
+        yError() << "[OculusModule::updateModule] Unable to move the right finger";
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/OculusModule.cpp]: Maybe we can avoid to return ``false`` if we are not able to move some finger.

File: app/robots/iCubGenova04/#leftFingersRetargetingParams.ini# [hunk 10]
```diff
@@ -0,0 +1,4 @@
+remote_control_boards   ("left_arm")
```
[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply app/robots/iCubGenova04/#leftFingersRetargetingParams.ini#]: Yep

File: modules/Oculus_module/include/HeadRetargeting.hpp [hunk 11]
```diff
@@ -19,27 +19,45 @@
...
+// iDynTree
+#include <iDynTree/Core/Rotation.h>
+
+#include <RetargetingHelper.hpp>
+
 /**
  * Class useful to manage the head retargeting.
  */
-class HeadRetargeting
+class HeadRetargeting : public RetargetingHelper
```
[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply modules/Oculus_module/include/HeadRetargeting.hpp]: Yes, you are right. Do you have any suggestion?

File: modules/Oculus_module/src/RobotControlHelper.cpp (hunk 7 above)
Comment (GiulioRomualdi) [reply modules/Oculus_module/src/RobotControlHelper.cpp]: Yes, I can move the `getLimts()` inside the `RobotControlHelper::getLimits()`. Indeed the limits are not used by the `RobotControlHelper` class itself (except [here](https://github.com/GiulioRomualdi/walking-teleoperation/blob/refactoring/oculus_module/modules/Oculus_module/src/RobotControlHelper.cpp#L197-L201) but as discussed above we may remove those lines)

File: modules/Oculus_module/src/RobotControlHelper.cpp (hunk 8 above)
Comment (GiulioRomualdi) [reply modules/Oculus_module/src/RobotControlHelper.cpp]: Instead of zero, the method may set the limits equal to the current encoder value

File: app/robots/iCubGenova04/.#leftFingersRetargetingParams.ini (hunk 1 above)
Comment (GiulioRomualdi) [reply app/robots/iCubGenova04/.#leftFingersRetargetingParams.ini]: Done in abaa3ee

File: modules/Oculus_module/include/HeadRetargeting.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply modules/Oculus_module/include/HeadRetargeting.hpp]: What do you think about ``RetargetingController``?

[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply modules/Oculus_module/include/HeadRetargeting.hpp]: Done in 0cc66dc

File: modules/Oculus_module/src/RobotControlHelper.cpp (hunk 7 above)
Comment (GiulioRomualdi) [reply modules/Oculus_module/src/RobotControlHelper.cpp]: Done in 185a6dd

File: modules/Oculus_module/src/RobotControlHelper.cpp (hunk 8 above)
Comment (GiulioRomualdi) [reply modules/Oculus_module/src/RobotControlHelper.cpp]: Done in 185a6dd

File: modules/Oculus_module/src/RobotControlHelper.cpp (hunk 6 above)
Comment (GiulioRomualdi) [reply modules/Oculus_module/src/RobotControlHelper.cpp]: Done in 185a6dd

File: modules/Oculus_module/src/HeadRetargeting.cpp [hunk 12]
```diff
@@ -39,29 +93,37 @@ bool HeadRetargeting::configure(const yarp::os::Searchable& config)
...
 {
-    yarp::sig::Vector desiredHeadOrientation;
-    desiredHeadOrientation = m_desiredHeadOrientation;
-    desiredHeadOrientation(2) = desiredHeadOrientation(2) + m_playerOrientation;
+    m_desiredHeadOrientation = m_playerOrientation.inverse() * m_oculusRoot_T_oculusHeadset;
 
-    m_headTrajectorySmoother->computeNextValues(desiredHeadOrientation);
-}
+    yarp::sig::Vector tmp(3);
+    iDynTree::toYarp(inverseKinematics(m_desiredHeadOrientation), tmp);
```
[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply modules/Oculus_module/src/HeadRetargeting.cpp]: Done in e35a78b

File: modules/Oculus_module/src/HeadRetargeting.cpp [hunk 13]
```diff
@@ -7,10 +7,57 @@
...
+
+    // minus due to the joints structure of the iCub neck
+    YXZ(0) = -thetaY;
+    YXZ(1) = thetaX;
+    YXZ(2) = thetaZ;
+
+    return YXZ;
+}
+
+iDynTree::Rotation HeadRetargeting::forwardKinematics(const yarp::sig::Vector& YXZ)
```
[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply modules/Oculus_module/src/HeadRetargeting.cpp]: Done in e35a78b

File: modules/Oculus_module/src/RobotControlHelper.cpp [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+
+    options.put("remoteControlBoards", remoteControlBoards.get(0));
+    options.put("localPortPrefix", "/" + name + "/remoteControlBoard");
+    yarp::os::Property& remoteControlBoardsOpts = options.addGroup("REMOTE_CONTROLBOARD_OPTIONS");
+    remoteControlBoardsOpts.put("writeStrict", "on");
+
+    m_actuatedDOFs = m_axesList.size();
+
+    // open the device
+    if (!m_robotDevice.open(options) && m_isMandatory)
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/RobotControlHelper.cpp]: Notice that in this way it returns false every time ``m_isMandatory`` is false. Anyway if the ``open`` returns ``false`` is always worthwhile to throw an error.

Comment (S-Dafarra) [reply modules/Oculus_module/src/RobotControlHelper.cpp]: Is this the call failing when some finger is in ``NOT CONFIGURED`` status?

File: modules/Oculus_module/src/RobotControlHelper.cpp [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+        yError() << "[RobotControlHelper::getLimits] Unable to read encoders (position).";
+        return false;
+    }
+
+    // resize matrix
+    limits.resize(m_actuatedDOFs, 2);
+
+    for (int i = 0; i < m_actuatedDOFs; i++)
+        // get position limits
+        if (m_limitsInterface->getLimits(i, &limits(i, 0), &limits(i, 0)))
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/RobotControlHelper.cpp]: Setting twice the lower bound.

File: app/robots/iCubGenova04/headRetargetingParams.ini (hunk 2 above)
Comment (S-Dafarra) [reply app/robots/iCubGenova04/headRetargetingParams.ini]: Rather than "not wrong", this specific order is necessary to have the methods ``forward`` and ``inverse kinematics`` correctly working as they are used right now.

**Comment by GiulioRomualdi** [comment]
The code was tested on the real robot and the documentation was improved. When you have time, can you finish reviewing the PR? I'm also trying to add @kouroshD as a reviewer but seems that GitHub is not able to find him.

**Comment by DanielePucci** [comment]
CC
@S-Dafarra @traversaro please take note of the kind above request by @GiulioRomualdi . @kouroshD added into this repo with *admin* rights and as reviewer of this PR

**Comment by traversaro** [comment]
> The code was tested on the real robot and the documentation was improved. When you have time, can you finish reviewing the PR?

Sorry, I lost this request in the sea of commits.

File: modules/Oculus_module/src/HeadRetargeting.cpp [hunk 16]
```diff
@@ -7,30 +7,81 @@
...
+#include <iDynTree/yarp/YARPEigenConversions.h>
 
-bool HeadRetargeting::configure(const yarp::os::Searchable& config)
+#include <HeadRetargeting.hpp>
+#include <Utils.hpp>
+
+void HeadRetargeting::inverseKinematics(const iDynTree::Rotation& matrix, double& neckPitch,
+                                        double& neckRoll, double& neckYaw)
+{
+    // YXZ decomposition
```
Comment (traversaro) [inline modules/Oculus_module/src/HeadRetargeting.cpp]: If you have a reference (paper/book/other code) on this code, feel free to add it.

File: modules/Oculus_module/src/HeadRetargeting.cpp [hunk 17]
```diff
@@ -7,30 +7,81 @@
...
+    if (matrix(1, 2) < 1)
+    {
+        if (matrix(1, 2) > -1)
+        {
+            neckRoll = std::asin(-matrix(1, 2));
+            neckPitch = std::atan2(matrix(0, 2), matrix(2, 2));
+            neckYaw = std::atan2(matrix(1, 0), matrix(1, 1));
+        } else
+        {
+            neckRoll = M_PI / 2;
```
Comment (traversaro) [inline modules/Oculus_module/src/HeadRetargeting.cpp]: You use `M_PI`, but this will create a failure in compiling in Windows because (as far as I can see) `_USE_MATH_DEFINES` is not defined for this compilation unit, see https://github.com/robotology/QA/issues/81 . 
I suggest  to either add `target_compile_definitions(${EXE_TARGET_NAME} PRIVATE _USE_MATH_DEFINES)`, or even better avoid the use of M_PI.

**Comment by traversaro** [comment]
In general, I think that one reviewer per PR is enough, so as long as you have one reviewer that is approving the PR, it make sense to merge the PR.

File: app/robots/iCubGenova04/leftHandRetargetingParams.ini [hunk 18]
```diff
@@ -1 +1,14 @@
...
+handOculusFrame_R_handRobotFrame ((1.0 0.0 0.0), (0.0 1.0 0.0), (0.0 0.0 1.0))
+
+# the following rotation map the teleoperation frame and the teleoperation robot
+# frame in this case the teleoperation robot frame chosen is "imu_frame" and
+# according to icub cad has:
+# The z-axis is parallel to gravity but pointing upwards.
+# The x-axis points behind the robot.
+# The y-axis points laterally and is chosen according to the right-hand rule.
+# So the rotation matrix is simply Rotz(pi)
+teleoperationRobotFrame_R_teleoperationFrame ((-1.0 0.0 0.0), (0.0 -1.0 0.0), (0.0 0.0 1.0))
```
Comment (S-Dafarra) [inline app/robots/iCubGenova04/leftHandRetargetingParams.ini]: Even if it can be devised from the above comments, it may be worthwhile to specify how the teleoperation frame is oriented.

File: modules/Oculus_module/include/HeadRetargeting.hpp [hunk 19]
```diff
@@ -19,27 +19,81 @@
...
+    void setDesiredHeadOrientation(const yarp::sig::Matrix& oculusInertial_T_headOculus);
+
+    /**
+     * Evaluate the inverse kinematics of the head
+     * @param matrix is the rotation matrix of the head with respect the root frame
+     * @param neckPitch neck pitch angle expressed in radiant
+     * @param neckRoll neck roll angle expressed in radiant
+     * @param neckYaw neck yaw angle expressed in radiant
+     */
+    static void inverseKinematics(const iDynTree::Rotation& matrix, double& neckPitch,
```
[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply modules/Oculus_module/include/HeadRetargeting.hpp]: Actually, I didn't understand. In the code the joints are named `neckPitch`, `neckRoll` and `neckYaw` while in the [iCubWiki](http://wiki.icub.org/wiki/ICub_joints) the only difference is that  `underscore_case` is used rather than the `cammelCase`.

File: modules/Oculus_module/src/HeadRetargeting.cpp (hunk 17 above)
Comment (GiulioRomualdi) [reply modules/Oculus_module/src/HeadRetargeting.cpp]: I could use `iDynTree::rad2deg(90)` what do you think?

Comment (GiulioRomualdi) [reply modules/Oculus_module/src/HeadRetargeting.cpp]: Ops :smile:

Comment (GiulioRomualdi) [reply modules/Oculus_module/src/HeadRetargeting.cpp]: Done in 50102ab

File: app/robots/iCubGenova04/leftHandRetargetingParams.ini (hunk 18 above)
Comment (GiulioRomualdi) [reply app/robots/iCubGenova04/leftHandRetargetingParams.ini]: Done in a19e44d

File: modules/Oculus_module/include/HeadRetargeting.hpp [hunk 20]
```diff
@@ -19,27 +19,81 @@
...
+    /**
+     * Set the desired head orientation.
+     * @param oculusInertial_T_headOculus is the homogeneous transformation between the oculus
+     * inertial frame and the head oculus frame
+     */
+    void setDesiredHeadOrientation(const yarp::sig::Matrix& oculusInertial_T_headOculus);
+
+    /**
+     * Evaluate the inverse kinematics of the head
+     * @param matrix is the rotation matrix of the head with respect the root frame
```
[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply modules/Oculus_module/include/HeadRetargeting.hpp]: Done in b88df10

**Comment by GiulioRomualdi** [comment]
If you do not have any other suggestions I merge the PR
